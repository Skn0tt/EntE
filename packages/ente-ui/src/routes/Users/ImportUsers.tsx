/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import { AppState, createUsersRequest, getStudents } from "../../redux";
import { Button, Dialog, Grid } from "@material-ui/core";
import withMobileDialog from "@material-ui/core/withMobileDialog";
import DialogActions from "@material-ui/core/DialogActions";
import * as React from "react";
import Dropzone from "react-dropzone";
import { connect, MapStateToPropsParam } from "react-redux";
import { Action, Dispatch } from "redux";
import { CreateUserDto } from "ente-types";
import { parseCSVFromFile } from "../../helpers/parser";
import SignedAvatar from "../../elements/SignedAvatar";
import { createTranslation } from "../../helpers/createTranslation";
import { MessagesContext, MessagesContextValue } from "../../context/Messages";

const lang = createTranslation({
  en: {
    submit: "Import",
    close: "Close",
    dropzone: "Drop a .csv file here."
  },
  de: {
    submit: "Importieren",
    close: "SchlieÃŸen",
    dropzone: "Legen sie hier eine .csv-Datei ab."
  }
});

const readFile = async (f: File) =>
  new Promise<string>((resolve, reject) => {
    const r = new FileReader();
    r.onload = _ => resolve(r.result as string);
    r.onerror = reject;
    r.readAsText(f);
  });

/**
 * # Component Types
 */
interface OwnProps {
  onClose(): void;
  show: boolean;
}

interface StateProps {
  usernames: string[];
}
const mapStateToProps: MapStateToPropsParam<
  StateProps,
  OwnProps,
  AppState
> = state => ({
  usernames: getStudents(state).map(u => u.get("username"))
});

interface DispatchProps {
  createUsers(users: CreateUserDto[]): void;
}
const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  createUsers: (users: CreateUserDto[]) => dispatch(createUsersRequest(users))
});

interface InjectedProps {
  fullScreen: boolean;
}

interface State {
  users: CreateUserDto[];
  error: boolean;
}

type Props = OwnProps & DispatchProps & StateProps;

/**
 * # Component
 */
export class ImportUsers extends React.Component<Props & InjectedProps, State> {
  /**
   * ## Intialization
   */
  state: State = {
    users: [],
    error: true
  };

  static contextType = MessagesContext;

  /**
   * # Handlers
   */
  handleClose = () => this.props.onClose();

  handleSubmit = () =>
    this.state.users.length !== 0 && this.props.createUsers(this.state.users);

  onDrop = async (accepted: File[], rejected: File[]) => {
    if (accepted.length === 0) {
      return;
    }

    try {
      const input = await readFile(accepted[0]);
      const users = await parseCSVFromFile(input, this.props.usernames);
      this.setState({ users, error: false });
    } catch (error) {
      this.setState({ error: true });
      const { addMessages } = this.context as MessagesContextValue;
      addMessages(error.message);
    }
  };

  /**
   * # Validation
   */
  inputValid = (): boolean =>
    !this.state.error && this.state.users.length !== 0;

  /**
   * # Render
   */
  render() {
    const { show, fullScreen } = this.props;
    const { error } = this.state;

    return (
      <Dialog fullScreen={fullScreen} onClose={this.handleClose} open={show}>
        <Grid container direction="column">
          <Grid item xs={12}>
            <Dropzone
              accept="text/csv"
              onDrop={this.onDrop}
              className="dropzone"
            >
              {lang.dropzone}
            </Dropzone>
          </Grid>
          <Grid item xs={12}>
            <SignedAvatar signed={!error} />
          </Grid>
        </Grid>
        <DialogActions>
          <Button
            onClick={this.handleClose}
            color="secondary"
            className="close"
          >
            {lang.close}
          </Button>
          <Button
            onClick={() => {
              this.handleSubmit();
              this.handleClose();
            }}
            disabled={!this.inputValid()}
            color="primary"
            className="submit"
          >
            {lang.submit}
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

export default connect<StateProps, DispatchProps, OwnProps, AppState>(
  mapStateToProps,
  mapDispatchToProps
)(withMobileDialog<Props>()(ImportUsers));
