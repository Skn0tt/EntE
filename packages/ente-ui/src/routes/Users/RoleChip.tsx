import * as React from "react";
import { Roles } from "ente-types";
import { Chip, Theme } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { RoleTranslation } from "../../roles.translation";

const useStyles = makeStyles((theme: Theme) => ({
  [RoleTranslation.de.admin]: {
    backgroundColor: "rgb(222, 50, 21)",
    color: theme.palette.common.white
  },
  [RoleTranslation.de.manager]: {
    backgroundColor: "#f57c00",
    color: theme.palette.common.white
  },
  [RoleTranslation.de.parent]: {
    backgroundColor: "rgb(14, 115, 198)",
    color: theme.palette.common.white
  },
  [RoleTranslation.de.student]: {
    backgroundColor: "rgb(104, 159, 56)",
    color: theme.palette.common.white
  },
  [RoleTranslation.de.teacher]: {
    backgroundColor: "rgb(69, 90, 100)",
    color: theme.palette.common.white
  },
  [RoleTranslation.en.admin]: {
    backgroundColor: "rgb(222, 50, 21)",
    color: theme.palette.common.white
  },
  [RoleTranslation.en.manager]: {
    backgroundColor: "#f57c00",
    color: theme.palette.common.white
  },
  [RoleTranslation.en.parent]: {
    backgroundColor: "rgb(14, 115, 198)",
    color: theme.palette.common.white
  },
  [RoleTranslation.en.student]: {
    backgroundColor: "rgb(104, 159, 56)",
    color: theme.palette.common.white
  },
  [RoleTranslation.en.teacher]: {
    backgroundColor: "rgb(69, 90, 100)",
    color: theme.palette.common.white
  }
}));

interface RoleChipProps {
  translatedRole: string;
}

export const RoleChip: React.FC<RoleChipProps> = props => {
  const { translatedRole } = props;

  const classes = useStyles();

  return (
    <Chip
      label={translatedRole}
      classes={{
        root: classes[translatedRole]
      }}
    />
  );
};
