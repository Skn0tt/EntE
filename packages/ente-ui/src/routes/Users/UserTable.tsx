import * as React from "react";
import Table from "../../components/Table";
import { UserN } from "../../redux";
import { createTranslation } from "../../helpers/createTranslation";

const lang = createTranslation({
  en: {
    headers: {
      username: "Username",
      displayname: "Displayname",
      email: "Email",
      role: "Role"
    }
  },
  de: {
    headers: {
      username: "Nutzername",
      displayname: "Displayname",
      email: "Email",
      role: "Rolle"
    }
  }
});

interface UserTableOwnProps {
  users: UserN[];
  onClick?: (id: string) => void;
}

class _UserTable extends Table<UserN> {}

export const UserTable: React.SFC<UserTableOwnProps> = props => {
  const { users, onClick } = props;

  return (
    <_UserTable
      headers={[
        {
          name: lang.headers.username,
          options: {
            filter: false
          }
        },
        {
          name: lang.headers.displayname,
          options: {
            filter: false
          }
        },
        {
          name: lang.headers.email,
          options: {
            filter: false
          }
        },
        {
          name: lang.headers.role
        }
      ]}
      items={users}
      extract={user => [
        user.get("username")!,
        user.get("displayname")!,
        user.get("email")!,
        user.get("role")!
      ]}
      extractId={user => user.get("id")}
      onClick={onClick}
    />
  );
};
