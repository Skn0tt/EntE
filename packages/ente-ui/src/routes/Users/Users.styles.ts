/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import { StyleRules } from "@material-ui/core/styles/withStyles";
import { Theme } from "@material-ui/core/styles/createMuiTheme";

const styles = (theme: Theme): StyleRules => ({
  searchBar: {
    padding: 10
  },
  fab: {
    margin: 0,
    top: "auto",
    right: theme.spacing(2),
    bottom: theme.spacing(2),
    left: "auto",
    position: "fixed"
  },
  table: {
    overflowX: "auto"
  }
});

export default styles;
