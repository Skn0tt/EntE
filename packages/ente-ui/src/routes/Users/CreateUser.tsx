/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import {
  connect,
  MapStateToPropsParam,
  MapDispatchToPropsParam
} from "react-redux";

import { Dialog, Button, Grid, TextField, Typography } from "@material-ui/core";
import DialogTitle from "@material-ui/core/DialogTitle";
import DialogContent from "@material-ui/core/DialogContent";
import DialogActions from "@material-ui/core/DialogActions";
import withMobileDialog, {
  InjectedProps
} from "@material-ui/core/withMobileDialog";
import { Action } from "redux";
import ImportUsers from "./ImportUsers";
import {
  Roles,
  CreateUserDto,
  isValidDisplayname,
  isValidUsername,
  isValidEmail,
  isValidPassword,
  createDefaultCreateUserDto,
  isValidCreateUserDto,
  roleHasBirthday
} from "ente-types";
import {
  getStudents,
  getUser,
  createUsersRequest,
  AppState,
  UserN,
  roleHasChildren,
  roleHasGradYear
} from "../../redux";
import TextInput from "../../elements/TextInput";
import ChildrenInput from "../../elements/ChildrenInput";
import withErrorBoundary from "../../hocs/withErrorBoundary";
import { YearPicker } from "../../elements/YearPicker";
import { PasswordRequirementsHint } from "../../elements/PasswordRequirementsHint";
import * as _ from "lodash";
import { Maybe } from "monet";
import {
  WithTranslation,
  withTranslation
} from "../../helpers/with-translation";
import { DateInput } from "../../elements/DateInput";

export const lang = {
  en: {
    titles: {
      newUser: "New User",
      username: "Username",
      displayname: "Displayname",
      email: "Email",
      password: "Password",
      role: "Role",
      gradYear: "Graduation Year",
      birthday: "Birthday"
    },
    helpers: {
      chooseRoleOfUser: "Choose the users's role"
    },
    submit: "Submit",
    close: "Close",
    import: "Import",
    passwordSpec: PasswordRequirementsHint.en
  },
  de: {
    titles: {
      newUser: "Neuer Nutzer",
      username: "Nutzername",
      displayname: "Displayname",
      email: "Email",
      password: "Passwort",
      role: "Rolle",
      gradYear: "Abschluss-Jahrgang",
      birthday: "Geburtstag"
    },
    helpers: {
      chooseRoleOfUser: "Wählen Sie die Rolle des Nutzers aus"
    },
    submit: "Erstellen",
    close: "Schließen",
    import: "Importieren",
    passwordSpec: PasswordRequirementsHint.de
  }
};

/**
 * # Component Types
 */
interface OwnProps {
  onClose(): void;
  show: boolean;
}

interface CreateUserState {
  create: CreateUserDto;
  showImportUsers: boolean;
}

interface CreateUserStateProps {
  students: UserN[];
  getUser(id: string): Maybe<UserN>;
}
const mapStateToProps: MapStateToPropsParam<
  CreateUserStateProps,
  OwnProps,
  AppState
> = state => ({
  getUser: id => getUser(id)(state),
  students: getStudents(state)
});

interface CreateUserDispatchProps {
  createUsers(...users: CreateUserDto[]): Action;
}
const mapDispatchToProps: MapDispatchToPropsParam<
  CreateUserDispatchProps,
  OwnProps
> = dispatch => ({
  createUsers: (...users: CreateUserDto[]) =>
    dispatch(createUsersRequest(users))
});

type CreateUserProps = OwnProps &
  CreateUserStateProps &
  CreateUserDispatchProps &
  WithTranslation<typeof lang.en> &
  InjectedProps;

/**
 * # Component
 */
export class CreateUser extends React.Component<
  CreateUserProps,
  CreateUserState
> {
  /**
   * # Intialization
   */
  state: Readonly<CreateUserState> = {
    create: createDefaultCreateUserDto(),
    showImportUsers: false
  };

  /**
   * ## Action Handlers
   */
  handleGoBack = () => this.handleClose();

  handleClose = () => this.props.onClose();

  handleShowImport = () => this.setState({ showImportUsers: true });
  handleCloseImport = () => this.setState({ showImportUsers: false });

  handleSubmit = () => {
    return this.props.createUsers(this.state.create);
  };

  handleKeyPress: React.KeyboardEventHandler<{}> = event => {
    if (event.key === "Enter" && this.inputValid()) {
      this.handleSubmit();
    }
  };

  handleImport = (u: CreateUserDto[]) => {
    this.props.createUsers(...u);
    this.handleCloseImport();
    this.handleClose();
  };

  /**
   * ## Input Handlers
   */
  update = (key: keyof CreateUserDto) => (value: any) => {
    const clone = Object.assign({}, this.state.create);
    clone[key] = value;
    this.setState({ create: clone });
  };
  handleChangeUsername = this.update("username");
  handleChangeYear = this.update("graduationYear");
  handleChangeDisplayname = this.update("displayname");
  handleChangePassword = (value: string) => {
    const newValue = value === "" ? undefined : value;
    this.update("password")(newValue);
  };
  handleChangeBirthday = (birthday: string) =>
    this.update("birthday")(birthday);
  handleChangeEmail = this.update("email");
  handleChangeRole = (event: React.ChangeEvent<HTMLInputElement>) =>
    this.update("role")(event.target.value as Roles);
  handleChangeChildren = this.update("children");

  hasChildren = (): boolean => {
    const {
      create: { role }
    } = this.state;
    return roleHasChildren(role);
  };

  hasGradYear = (): boolean => {
    const {
      create: { role }
    } = this.state;
    return roleHasGradYear(role);
  };

  hasBirthday = (): boolean => {
    const {
      create: { role }
    } = this.state;
    return roleHasBirthday(role);
  };

  /**
   * ## Validation
   */
  usernameValid = (): boolean => isValidUsername(this.state.create.username);
  displaynameValid = (): boolean =>
    isValidDisplayname(this.state.create.displayname);
  emailValid = (): boolean => isValidEmail(this.state.create.email);
  passwordValid = (): boolean =>
    _.isUndefined(this.state.create.password) ||
    isValidPassword(this.state.create.password);
  childrenValid = (): boolean =>
    !this.hasChildren() || this.state.create.children.length > 0;
  inputValid = (): boolean => isValidCreateUserDto(this.state.create);

  render() {
    const { show, fullScreen, getUser, students, translation } = this.props;
    const { create, showImportUsers } = this.state;

    return (
      <>
        <Dialog fullScreen={fullScreen} onClose={this.handleGoBack} open={show}>
          <DialogTitle>{translation.titles.newUser}</DialogTitle>
          <DialogContent>
            <form onKeyPress={this.handleKeyPress}>
              <Grid container direction="row">
                <Grid item xs={12} lg={6}>
                  <TextInput
                    value={create.username || ""}
                    label={translation.titles.username}
                    onChange={this.handleChangeUsername}
                    validator={this.usernameValid}
                    required
                  />
                </Grid>
                <Grid item xs={12} lg={6}>
                  <TextInput
                    label={translation.titles.displayname}
                    value={create.displayname || ""}
                    onChange={this.handleChangeDisplayname}
                    validator={this.displaynameValid}
                    required
                  />
                </Grid>
                <Grid item xs={12} lg={6}>
                  <TextInput
                    label={translation.titles.email}
                    value={create.email || ""}
                    onChange={this.handleChangeEmail}
                    validator={this.emailValid}
                    type="email"
                    required
                  />
                </Grid>
                <Grid item xs={12}>
                  <TextInput
                    label={translation.titles.password}
                    value={create.password || ""}
                    onChange={this.handleChangePassword}
                    validator={this.passwordValid}
                    type="password"
                  />
                </Grid>
                <Typography variant="body1">
                  <translation.passwordSpec />
                </Typography>
                <Grid item xs={12}>
                  <TextField
                    select
                    label={translation.titles.role}
                    value={create.role || Roles.STUDENT}
                    onChange={this.handleChangeRole}
                    fullWidth
                    SelectProps={{ native: true }}
                    helperText={translation.helpers.chooseRoleOfUser}
                  >
                    {Object.keys(Roles).map(role => (
                      <option
                        key={(Roles as any)[role]}
                        value={(Roles as any)[role]}
                      >
                        {(Roles as any)[role]}
                      </option>
                    ))}
                  </TextField>
                </Grid>
                {this.hasChildren() && (
                  <Grid item xs={12}>
                    <ChildrenInput
                      children={create.children.map(getUser).map(c => c.some())}
                      students={students}
                      onChange={(u: UserN[]) =>
                        this.handleChangeChildren(u.map(u => u.get("id")))
                      }
                    />
                  </Grid>
                )}
                {this.hasGradYear() && (
                  <Grid item xs={12}>
                    <YearPicker
                      label={translation.titles.gradYear}
                      amount={5}
                      onChange={this.handleChangeYear}
                      value={create.graduationYear!}
                    />
                  </Grid>
                )}
                {this.hasBirthday() && (
                  <Grid item xs={12}>
                    <DateInput
                      value={create.birthday!}
                      isValid={_ => true}
                      label={translation.titles.birthday}
                      onChange={this.handleChangeBirthday}
                      maxDate={Date.now()}
                    />
                  </Grid>
                )}
              </Grid>
            </form>
          </DialogContent>
          <DialogActions>
            <Button onClick={this.handleShowImport}>
              {translation.import}
            </Button>
            <Button
              onClick={this.handleClose}
              color="secondary"
              className="close"
            >
              {translation.close}
            </Button>
            <Button
              onClick={() => {
                this.handleSubmit();
                this.handleClose();
              }}
              disabled={!this.inputValid()}
              color="primary"
            >
              {translation.submit}
            </Button>
          </DialogActions>
        </Dialog>
        <ImportUsers
          onClose={this.handleCloseImport}
          show={showImportUsers}
          onImport={this.handleImport}
        />
      </>
    );
  }
}

export default connect<
  CreateUserStateProps,
  CreateUserDispatchProps,
  OwnProps,
  AppState
>(
  mapStateToProps,
  mapDispatchToProps
)(
  withTranslation(lang)(
    withMobileDialog<CreateUserProps>()(withErrorBoundary()(CreateUser))
  )
);
