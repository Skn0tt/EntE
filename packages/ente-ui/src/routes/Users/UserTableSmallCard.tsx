import * as React from "react";
import { UserN } from "../../redux";
import { Card, CardContent, Typography, Theme } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { RoleChip } from "./RoleChip";
import { useRoleTranslation } from "../../roles.translation";

const useStyles = makeStyles((theme: Theme) => ({
  roleDiv: {
    float: "right",
    top: 0,
    right: 0
  },
  card: {
    margin: theme.spacing(1)
  }
}));

interface UserTableSmallCardProps {
  user: UserN;
  onClick: (user: UserN) => void;
}

export const UserTableSmallCard: React.FC<UserTableSmallCardProps> = props => {
  const { user, onClick } = props;

  const roleTranslation = useRoleTranslation();
  const classes = useStyles();

  const handleClick = React.useCallback(
    () => {
      onClick(user);
    },
    [onClick, user]
  );

  return (
    <Card onClick={handleClick} className={classes.card} elevation={1}>
      <CardContent>
        <div className={classes.roleDiv}>
          <RoleChip translatedRole={roleTranslation[user.get("role")]} />
        </div>

        <Typography color="textSecondary">{user.get("username")}</Typography>

        <Typography variant="h6">{user.get("displayname")}</Typography>

        <Typography component="p">{user.get("email")}</Typography>
      </CardContent>
    </Card>
  );
};
