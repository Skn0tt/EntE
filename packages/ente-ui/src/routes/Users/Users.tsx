/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import withStyles, { WithStyles } from "@material-ui/core/styles/withStyles";
import { connect, Dispatch } from "react-redux";
import styles from "./Users.styles";
import { Action } from "redux";
import CreateUser from "./CreateUser";
import { Button } from "@material-ui/core";
import { Add as AddIcon } from "@material-ui/icons";
import { AppState, getUsers, getUsersRequest, UserN } from "../../redux";
import { withRouter, RouteComponentProps } from "react-router";
import Table from "../../components/Table";
import withErrorBoundary from "../../hocs/withErrorBoundary";
import { createTranslation } from "ente-ui/src/helpers/createTranslation";

const lang = createTranslation({
  en: {
    headers: {
      username: "Username",
      displayname: "Displayname",
      email: "Email",
      role: "Role"
    }
  },
  de: {
    headers: {
      username: "Nutzername",
      displayname: "Displayname",
      email: "Email",
      role: "Rolle"
    }
  }
});

class UserTable extends Table<UserN> {}

/**
 * # Component Types
 */
interface StateProps {
  users: UserN[];
}
const mapStateToProps = (state: AppState) => ({
  users: getUsers(state)
});

interface DispatchProps {
  getUsers(): Action;
}
const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  getUsers: () => dispatch(getUsersRequest())
});

type UsersProps = StateProps &
  DispatchProps &
  WithStyles &
  RouteComponentProps<{}>;

interface UsersState {
  showCreateModal: boolean;
}

/**
 * # Component
 */
export class Users extends React.PureComponent<UsersProps, UsersState> {
  state: UsersState = {
    showCreateModal: false
  };

  componentDidMount() {
    this.props.getUsers();
  }

  showCreateModal = () => this.setState({ showCreateModal: true });
  closeCreateModal = () => this.setState({ showCreateModal: false });

  render() {
    const { users, history, classes } = this.props;

    return (
      <React.Fragment>
        {/* Modals */}
        <CreateUser
          onClose={this.closeCreateModal}
          show={this.state.showCreateModal}
        />

        {/* Main */}
        <UserTable
          headers={[
            {
              name: lang.headers.username,
              options: {
                filter: false
              }
            },
            {
              name: lang.headers.displayname,
              options: {
                filter: false
              }
            },
            {
              name: lang.headers.email,
              options: {
                filter: false
              }
            },
            {
              name: lang.headers.role
            }
          ]}
          items={users}
          extract={user => [
            user.get("username")!,
            user.get("displayname")!,
            user.get("email")!,
            user.get("role")!
          ]}
          extractId={user => user.get("id")}
          onClick={id => history.push(`/users/${id}`)}
        />

        {/* FAB */}
        <Button
          color="primary"
          variant="fab"
          onClick={this.showCreateModal}
          className={classes.fab}
        >
          <AddIcon />
        </Button>
      </React.Fragment>
    );
  }
}

export default withStyles(styles)(
  connect(mapStateToProps, mapDispatchToProps)(
    withRouter(withErrorBoundary()(Users))
  )
);
