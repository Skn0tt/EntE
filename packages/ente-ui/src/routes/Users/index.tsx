/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import withStyles, { WithStyles } from "material-ui/styles/withStyles";
import { connect, Dispatch } from "react-redux";
import styles from "./styles";
import { Action } from "redux";
import CreateUser from "./components/CreateUser";
import { Button } from "material-ui";
import { Add as AddIcon } from "material-ui-icons";
import { User, AppState, getUsers, getUsersRequest } from "ente-redux";
import withErrorBoundary from "../../components/withErrorBoundary";
import Table from "../../components/Table";
import { withRouter, RouteComponentProps } from "react-router";

class UserTable extends Table<User> {}

/**
 * # Component Types
 */
interface StateProps {
  users: User[];
}
const mapStateToProps = (state: AppState) => ({
  users: getUsers(state)
});

interface DispatchProps {
  getUsers(): Action;
}
const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  getUsers: () => dispatch(getUsersRequest())
});

type Props = StateProps & DispatchProps & WithStyles & RouteComponentProps<{}>;

interface State {
  showCreateModal: boolean;
}

/**
 * # Component
 */
export class Users extends React.PureComponent<Props, State> {
  state: State = {
    showCreateModal: false
  };

  componentDidMount() {
    this.props.getUsers();
  }

  showCreateModal = () => this.setState({ showCreateModal: true });
  closeCreateModal = () => this.setState({ showCreateModal: false });

  render() {
    const { users, history, classes } = this.props;

    return (
      <React.Fragment>
        {/* Modals */}
        <CreateUser
          onClose={this.closeCreateModal}
          show={this.state.showCreateModal}
        />

        {/* Main */}
        <UserTable
          headers={[
            {
              name: "Username",
              options: {
                filter: false
              }
            },
            {
              name: "Displayname",
              options: {
                filter: false
              }
            },
            {
              name: "Email",
              options: {
                filter: false
              }
            },
            {
              name: "Role"
            }
          ]}
          items={users}
          extract={user => [
            user.get("username")!,
            user.get("displayname")!,
            user.get("email")!,
            user.get("role")!
          ]}
          extractId={user => user.get("_id")}
          onClick={id => history.push(`/users/${id}`)}
        />

        {/* FAB */}
        <Button
          color="primary"
          variant="fab"
          onClick={this.showCreateModal}
          className={classes.fab}
        >
          <AddIcon />
        </Button>
      </React.Fragment>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(
  withStyles(styles)(withRouter(withErrorBoundary()(Users)))
);
