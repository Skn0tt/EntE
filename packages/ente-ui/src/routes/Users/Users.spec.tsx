/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import { Users } from "./Users";
import { shallow } from "enzyme";
import { getMockRouterProps } from "../../../testHelpers/mockRouter";
import { Roles } from "ente-types";
import { UserN } from "../../redux";

describe("Users", () => {
  const users: UserN[] = [
    new UserN({
      username: "simon",
      displayname: "Simon",
      email: "email@emai.com",
      birthday: "2100-01-01",
      role: Roles.STUDENT,
      id: "mystupidid"
    })
  ];
  const getUsers = jest.fn();
  const comp = shallow(
    <Users
      {...getMockRouterProps({})}
      classes={{}}
      users={users}
      getUsers={getUsers}
    />
  );

  it("renders correctly", () => {
    expect(comp).toMatchSnapshot();
  });

  it("requests all users", () => {
    expect(getUsers).toBeCalled();
  });
});
