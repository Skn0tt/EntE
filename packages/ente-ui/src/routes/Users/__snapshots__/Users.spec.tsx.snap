// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Users
    classes={Object {}}
    getUsers={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": Object {},
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "key": "",
        "pathname": "",
        "search": "",
        "state": Object {},
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "",
        "url": "",
      }
    }
    staticContext={Object {}}
    users={
      Array [
        Immutable.UserN {
          "id": "mystupidid",
          "username": "simon",
          "children": Array [],
          "displayname": "Simon",
          "email": "email@emai.com",
          "isAdult": false,
          "role": "student",
          "childrenIds": Array [],
          "graduationYear": null,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(WithTheme(WithWidth(WithMobileDialog)))
          onClose={[Function]}
          show={false}
        />,
        <Unknown
          onClick={[Function]}
          users={
            Array [
              Immutable.UserN {
                "id": "mystupidid",
                "username": "simon",
                "children": Array [],
                "displayname": "Simon",
                "email": "email@emai.com",
                "isAdult": false,
                "role": "student",
                "childrenIds": Array [],
                "graduationYear": null,
              },
            ]
          }
        />,
        <WithStyles(Fab)
          color="primary"
          onClick={[Function]}
        >
          <pure(AddIcon) />
        </WithStyles(Fab)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onClose": [Function],
          "show": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onClick": [Function],
          "users": Array [
            Immutable.UserN {
              "id": "mystupidid",
              "username": "simon",
              "children": Array [],
              "displayname": "Simon",
              "email": "email@emai.com",
              "isAdult": false,
              "role": "student",
              "childrenIds": Array [],
              "graduationYear": null,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <pure(AddIcon) />,
          "className": undefined,
          "color": "primary",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(WithTheme(WithWidth(WithMobileDialog)))
            onClose={[Function]}
            show={false}
          />,
          <Unknown
            onClick={[Function]}
            users={
              Array [
                Immutable.UserN {
                  "id": "mystupidid",
                  "username": "simon",
                  "children": Array [],
                  "displayname": "Simon",
                  "email": "email@emai.com",
                  "isAdult": false,
                  "role": "student",
                  "childrenIds": Array [],
                  "graduationYear": null,
                },
              ]
            }
          />,
          <WithStyles(Fab)
            color="primary"
            onClick={[Function]}
          >
            <pure(AddIcon) />
          </WithStyles(Fab)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onClose": [Function],
            "show": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
            "users": Array [
              Immutable.UserN {
                "id": "mystupidid",
                "username": "simon",
                "children": Array [],
                "displayname": "Simon",
                "email": "email@emai.com",
                "isAdult": false,
                "role": "student",
                "childrenIds": Array [],
                "graduationYear": null,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <pure(AddIcon) />,
            "className": undefined,
            "color": "primary",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
